cpu/sam0_common/sam0_sdhc/sdhc.c

```diff
@@ -0,0 +1,1062 @@
+        start *= SD_MMC_BLOCK_SIZE;
+        end   *= SD_MMC_BLOCK_SIZE;
+    }
+
+    if (!_wait_not_busy(state)) {
+        res = -EBUSY;
+        goto out;
+    }
+
+    sdhc_send_cmd(state, SD_CMD32_ERASE_WR_BLK_START, start);
+    sdhc_send_cmd(state, SD_CMD33_ERASE_WR_BLK_END, end);
+    if (!sdhc_send_cmd(state, SDMMC_CMD38_ERASE, 0)) {
+        res = -EIO;
+    }
+
+out:
+    mutex_unlock(&state->lock);
+    return res;
+}
+
+/**
+ * @brief CMD13 - Addressed card sends its status register.
+ * This function waits the clear of the busy flag
+ *
+ * @return true if success, otherwise false
+ */
+static bool _wait_not_busy(sdhc_state_t *state)
+{
+    uint32_t timeout;
+
+    /* Wait for data ready status.
+     * Nec timing: 0 to unlimited
+     * However a timeout is used.
+     * 200 000 * 8 cycles
+     */
+    timeout = 200000;
+    do {
+        if (!sdhc_send_cmd(state, SDMMC_MCI_CMD13_SEND_STATUS, (uint32_t)state->rca << 16)) {
+            return false;
+        }
+        /* Check busy flag */
+        if (SDHC_DEV->RR[0].reg & CARD_STATUS_READY_FOR_DATA) {
+            break;
+        }
+        if (timeout-- == 0) {
+            return false;
+        }
+    } while (1);
+
+    return true;
+}
+
+/**
+ * @brief Try to get the SDIO card's operating condition
+ * - CMD5 to read OCR NF field
+ * - CMD5 to wait OCR power up busy
+ * - CMD5 to read OCR MP field
+ *   sd_mmc_card->type is updated
+ *
+ * @return true if success, otherwise false
+ */
+static bool sdio_test_type(sdhc_state_t *state)
+{
+    uint32_t response;
+    uint32_t cmd5_retry = 5000;
+
+    /* CMD5 - SDIO send operation condition (OCR) command. */
+    if (!sdhc_send_cmd(state, SDIO_CMD5_SEND_OP_COND, 0)) {
+        return true; /* No error but card type not updated */
+    }
+    response = SDHC_DEV->RR[0].reg;
+    if ((response & OCR_SDIO_NF) == 0) {
+        return true; /* No error but card type not updated */
+    }
+
+    /*
+     * Wait card ready
+     * Timeout 1s = 400KHz / ((6+4)*8) cylces = 5000 retry
+     * 6 = cmd byte size
+     * 4(SPI) 6(MCI) = response byte size
+     */
+    while (1) {
+        response &= OCR_VDD_27_28 | OCR_VDD_28_29 | OCR_VDD_29_30
+                  | OCR_VDD_30_31 | OCR_VDD_31_32 | OCR_VDD_32_33;
+        /* CMD5 - SDIO send operation condition (OCR) command.*/
+        if (!sdhc_send_cmd(state, SDIO_CMD5_SEND_OP_COND, response)) {
+            return false;
+        }
+        response = SDHC_DEV->RR[0].reg;
+        if ((response & OCR_POWER_UP_BUSY) == OCR_POWER_UP_BUSY) {
+            break;
+        }
+        if (cmd5_retry-- == 0) {
+            return false;
+        }
+    }
+    /* Update card type at the end of busy */
+    if ((response & OCR_SDIO_MP) > 0) {
+        state->type = CARD_TYPE_SD_COMBO;
+    }
+    else {
+        state->type = CARD_TYPE_SDIO;
+    }
+    return true;
+}
+
+static bool _test_high_speed(sdhc_state_t *state)
+{
+    uint8_t switch_status[SD_SW_STATUS_BSIZE] = { 0 };
```

Comment:
```suggestion
    alignas(uint32_t) uint8_t switch_status[SD_SW_STATUS_BSIZE] = { 0 };
```

Reply:
Why would this not be aligned? It's the first and only element in this function.

Reply:
Without requesting it from the compiler, the compiler is not required to align this. E.g. when inlining this function there might be stack memory to safe when not aligning it.